pygmt/clib/session.py

```diff
@@ -1801,29 +1798,32 @@ def virtualfile_in(  # noqa: PLR0912
                 warnings.warn(message=msg, category=RuntimeWarning, stacklevel=2)
             _data = (data,) if not isinstance(data, pathlib.PurePath) else (str(data),)
         elif kind == "vectors":
-            _data = [x, y]
-            if z is not None:
-                _data.append(z)
-            if extra_arrays:
-                _data.extend(extra_arrays)
-        elif kind == "matrix":  # turn 2-D arrays into list of vectors
-            if hasattr(data, "items") and not hasattr(data, "to_frame"):
+            if data is None:
+                # data is None, so data must be given via x/y/z.
+                _data = [x, y]
+                if z is not None:
+                    _data.append(z)
+                if extra_arrays:
+                    _data.extend(extra_arrays)
+            elif hasattr(data, "items") and not hasattr(data, "to_frame"):
                 # pandas.DataFrame or xarray.Dataset types.
                 # pandas.Series will be handled below like a 1-D numpy.ndarray.
                 _data = [array for _, array in data.items()]
-            elif hasattr(data, "ndim") and data.ndim == 2 and data.dtype.kind in "iuf":
-                # Just use virtualfile_from_matrix for 2-D numpy.ndarray
-                # which are signed integer (i), unsigned integer (u) or
-                # floating point (f) types
-                _virtualfile_from = self.virtualfile_from_matrix
-                _data = (data,)
             else:
                 # Python list, tuple, numpy.ndarray, and pandas.Series types
                 _data = np.atleast_2d(np.asanyarray(data).T)
+        elif kind == "matrix":
+            # GMT can only accept a 2-D matrix which are signed integer (i), unsigned
+            # integer (u) or floating point (f) types. For other data types, we need to
+            # use virtualfile_from_vectors instead, which turns the matrix into list of
+            # vectors and allows for better handling of string type inputs (e.g. for
+            # datetime data types).
+            _data = (data,)
+            if data.dtype.kind not in "iuf":
+                _virtualfile_from = self.virtualfile_from_vectors
```

Comment:
Missing test coverage for these lines?

Reply:
I can add a test for it. Before adding more tests, I'm wondering if we should split the big "test_clib_virtualfiles.py" file (with more than 500 lines) into separate smaller test files, i.e., one test file for each Session method. 

- `test_clib_virtualfile_in`
- `test_clib_virtualfile_from_vectors`
- `test_clib_virtualfile_from_matrix`
- `test_clib_open_virtualfile`

Reply:
We've split it before in #2784, so yes, ok to split it again :laughing:

Reply:
> We've split it before in #2784, so yes, ok to split it again 😆

Done in #3512.

Reply:
I've added a test in cfa32ed8e5558ac3215366949b504e83ad2fa4ca to cover this line.